### PR TITLE
[DXAA-651] ProdSec updates to onboarding tool

### PR DESCRIPTION
**Pull request by mchang16-auth**
### Changes

**Write guard & audit log (`env-credentials.ts`)**

- Added a 30 second write guard that prevents accidental double-writes to the same project. Overlapping writes are blocked with a descriptive error.
- Added an append-only audit log (`.auth0-mcp-writes.log`) recording each write with timestamp, framework, keys written, and target file.
- Both the write guard state file and audit log are automatically added to `.gitignore`.

**Dry-run mode (`env-credentials.ts`, `applications.ts`)**
- New `dry_run` parameter returns the computed key names without writing to disk, so callers can confirm what would be written before committing.

**Path validation hardening (`env-credentials.ts`, `quickstarts.ts`)**
- Replaced the loose absolute-path check with `validateProjectPath()`, which enforces that the resolved path stays within the current working directory and contains no `..` traversal sequences.
- Added Zod schema validation on envSnippet.fileName to reject paths with directory components at the CDN spec layer.
- Validated envSnippet.fileName again at use time before constructing the env file path.
- Error message for path traversal no longer reflects back user-supplied paths.

**Atomic writes & better parsing (`credentials-writer.ts`)**
- .env files are now written atomically via a temp file + rename(), preventing a half-written file on crash.
- Replaced the hand-rolled env parser with the `dotenv` library for correct handling of quoted values, escapes, and edge cases.
- `ensureGitignore` now returns a boolean and avoids duplicating the # Auth0 credentials comment header.
- Extracted `CREDENTIAL_FILE_MODE = 0o600` as a non-overridable constant; on POSIX, throws if the post-write stat mode doesn't match (previously chmod failures were silently swallowed)
- Rejects existing .env files exceeding 1 MB or that are not regular files (directories, symlinks, pipes) before reading; same guard applied to `parseEnvFile`

**Rate limiting**
- Enforces a 5-calls-per-60-second sliding window on `auth0_save_credentials_to_file` at the server dispatch level; returns a clear retry-after error on excess calls; other tools are unaffected

**CDN error handling (`env-credentials.ts`)**
- Supported frameworks that return a null spec (CDN failure, no cache) now return a clear network error instead of silently falling back to generic variable names.

**Env file write safeguards (`src/utils/env-credentials.ts`)**
- Detects when `project_path` looks like a web server document root (e.g. /var/www/..., or a directory named public, dist, build, static, etc.) and appends a warning to the success message advising to block .env file access in the web server config
- Has a deny list to deny writing of .env if in a system or home directory, or sensitive home directory (ex. `~/.ssh`)

**Schema (`src/tools/applications.ts`, `src/tools/onboarding.ts`, `src/tools/quickstarts.ts`)**
- Added `additionalProperties: false` to the input schemas for auth0_save_credentials_to_file, auth0_onboarding, and auth0_configure_and_get_quickstart_guide so safe guards from injection attacks and compliant MCP clients will reject calls with undocumented parameters (e.g. permissions: "640") 

### References
https://auth0team.atlassian.net/browse/DXAA-651
https://oktainc.atlassian.net/wiki/spaces/REX/pages/927957893/Security+Review+Report+-+PSREV-2889+Auth0+MCP+Server+auth0_save_credentials_to_file+tool+Updates

### Testing

New and expanded tests cover write guard logic, dry-run, atomic writes, path traversal rejection, .gitignore deduplication, parseEnvFile edge cases, and the CDN failure path.

- [X] This change adds unit test coverage
- [ ] This change adds integration test coverage

### Checklist

- [ ] I have read the [Auth0 general contribution guidelines](https://github.com/auth0/open-source-template/blob/master/GENERAL-CONTRIBUTING.md)
- [ ] I have read the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md)
- [ ] All existing and new tests complete without errors

### CODE REVIEW

File: src/utils/credentials-writer.ts [hunk 1]
```diff
@@ -140,16 +162,22 @@ function ensureGitignore(cwd: string, envFileName: string): void {
       // Check if the env file is already ignored
       if (!content.split('\n').some((line) => line.trim() === envFileName)) {
         log(`Adding ${envFileName} to .gitignore`);
-        fs.appendFileSync(gitignorePath, `\n# Auth0 credentials\n${envFileName}\n`, 'utf-8');
+        const hasAuthComment = content.includes('# Auth0 credentials');
+        const header = hasAuthComment ? '' : '\n# Auth0 credentials';
+        fs.appendFileSync(gitignorePath, `${header}\n${envFileName}\n`, 'utf-8');
+        return true;
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:189]: This basically avoids duplicate "Auth0 credentials" headers in the `.gitignore `file

File: src/utils/credentials-writer.ts [hunk 2]
```diff
@@ -161,12 +189,11 @@ function ensureGitignore(cwd: string, envFileName: string): void {
...
   if (!fs.existsSync(filePath)) return {};
-  const result: Record<string, string> = {};
-  for (const line of fs.readFileSync(filePath, 'utf-8').split('\n')) {
-    const match = line.match(/^([A-Za-z_][A-Za-z0-9_]*)=(.*)/);
-    if (match) result[match[1]] = match[2];
+  try {
+    return dotenvParse(fs.readFileSync(filePath, 'utf-8'));
+  } catch {
+    return {};
   }
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:219]: Here I replace the use of a regex with dotenv (as suggested in the security review)

File: src/utils/env-credentials.ts [hunk 3]
```diff
@@ -55,26 +57,32 @@ type ResolvedCredentials =
...
+  const resolved = path.resolve(projectPath);
+  const allowedRoot = process.cwd();
+  if (!resolved.startsWith(allowedRoot + path.sep) && resolved !== allowedRoot) {
+    return 'project_path must be within the current working directory';
+  }
+  if (!fs.existsSync(resolved) || !fs.statSync(resolved).isDirectory()) {
+    return 'project_path does not exist or is not a directory';
+  }
+  return null;
+}
```
Comment (mchang16-auth) [inline src/utils/env-credentials.ts:85]: Here on line 64, `path.resolve()` turns the path into an absolute path. We then check it to make sure exists and that it's within the current working directory

File: test/utils/credentials-writer.test.ts [hunk 4]
```diff
@@ -252,6 +316,28 @@ describe('credentials-writer', () => {
...
+
+      expect(parseEnvFile(envFilePath)).toEqual({ FOO: 'hello world', BAR: 'single' });
+    });
+
+    it('should ignore inline comments after unquoted values', () => {
+      fs.writeFileSync(envFilePath, 'FOO=bar # a comment\n', 'utf-8');
+
+      expect(parseEnvFile(envFilePath)).toEqual({ FOO: 'bar' });
+    });
   });
```
Comment (mchang16-auth) [inline test/utils/credentials-writer.test.ts:389]: These two test cases are due to the change from regex to dotenv library usage. Using dotenv will strip surrounding quotes and trims inline comments from unquoted values. Previously the regex would miss these cases

File: src/utils/env-credentials.ts [hunk 5]
```diff
@@ -312,3 +374,87 @@ async function buildFallbackCredentials(
...
   }
 }
+
+// ── Write guard ──────────────────────────────────────────────────────────────
+
+const WRITE_GUARD_FILE = '.auth0-mcp-state.json';
+// 2 minutes: long enough to catch rapid double-invocations in a multi-step AI
+// flow, short enough not to block an intentional re-run after a failed attempt.
+const WRITE_GUARD_WINDOW_MS = 2 * 60 * 1000;
+
```
Comment (mchang16-auth) [inline src/utils/env-credentials.ts:389]: QUESTION: Currently I have it set to not write if the same values were already written less than 2 minutes ago. It can be overrided if `force: true`, I'm thinking having the ability to override makes the guard meaningless. Do we think we should keep this feature?

File: src/utils/onboarding.ts [hunk 6]
```diff
@@ -1,5 +1,16 @@
...
 
+export const PROJECT_MARKERS = [
+  'package.json', 'Cargo.toml', 'pyproject.toml', 'go.mod',
+  'composer.json', 'pom.xml', 'build.gradle', 'build.gradle.kts', '.git',
+];
+
+export function hasProjectMarker(dir: string): boolean {
+  return PROJECT_MARKERS.some(marker => fs.existsSync(path.join(dir, marker)));
+}
+
```
Comment (mchang16-auth) [inline src/utils/onboarding.ts:37]: This is the third one from the Code Review section of the security review. I chose to implement the allow-list solution since it's OS independent and fits better for our use case of setting up apps

File: src/utils/credentials-writer.ts [hunk 7]
```diff
@@ -51,10 +54,14 @@ export async function writeCredentialsToEnv(
...
-        `Resolved path: "${resolvedPath}", allowed directory: "${allowedDir}"`
-    );
+    throw new Error('Security error: file path resolves outside the allowed directory');
+  }
+  // Symlink protection: if the target already exists, verify its dereferenced real path is also within allowedDir
+  if (fs.existsSync(resolvedPath)) {
+    const realPath = fs.realpathSync(resolvedPath);
+    if (!realPath.startsWith(allowedDir + path.sep) && realPath !== allowedDir) {
+      throw new Error('Security error: file path resolves outside the allowed directory');
+    }
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:72]: This covers the first case under the Code Review section, basically using fs.realpathSync(resolvedPath) to remove any issues with sym link

File: src/utils/env-credentials.ts (hunk 5 above)
Comment (wdaimee) [reply src/utils/env-credentials.ts:389]: I think it makes sense to have the guard in place. It prevents automatic abuse of writing to the `.env` file repeatedly; the LLM or some other script may try to write it repeatedly, but this guard prevents that. The user also sees a warning message saying that the environment variables were written less than XXX minutes ago, which is great. The user can manually force the tool to run. I think this is great from a security and UX perspectives, having the user manually force the tool to execute.

The only thing I would say is that the 2 mins seems like a long time. Can/should we reduce it? I am thinking roughly 30 seconds? What do you think?

File: src/utils/credentials-writer.ts [hunk 8]
```diff
@@ -84,27 +96,43 @@ export async function writeCredentialsToEnv(
     log(`Creating new file with credentials: ${envFile}`);
   }
 
-  fs.writeFileSync(envFile, content, 'utf-8');
+  const tmpFile = envFile + '.tmp';
+  try {
+    fs.writeFileSync(tmpFile, content, 'utf-8');
```
Comment (wdaimee) [inline src/utils/credentials-writer.ts]:
```suggestion
    fs.writeFileSync(tmpFile, content, { encoding: 'utf-8', mode: 0o600 });
```
There is a small window during the `.env` file creation where the secrets are vulnerable since we aren't setting `chmod 600` on the tmp file. We only set it on the final `.env` file created. Lets close this window by applying `chmod 600` to the tmp file when it is created.

File: src/utils/env-credentials.ts [hunk 9]
```diff
@@ -55,26 +57,37 @@ type ResolvedCredentials =
...
+  if (projectPath.includes('..')) {
+    return 'project_path must not contain path traversal sequences';
+  }
+  const resolved = path.resolve(projectPath);
+  const allowedRoot = process.cwd();
+
+  if (!fs.existsSync(resolved) || !fs.statSync(resolved).isDirectory()) {
+    return 'project_path does not exist or is not a directory';
+  }
+  if (!resolved.startsWith(allowedRoot + path.sep) && resolved !== allowedRoot) {
```
Comment (wdaimee) [inline src/utils/env-credentials.ts]: We need to remove this logic of only having it work in the current working directory. This will break implementation for MCP clients not in an IDE or in the project root. Claude Desktop, Windsurf, etc. stop working. We still want the tool to work in these cases.

File: src/utils/env-credentials.ts [hunk 10]
```diff
@@ -55,26 +57,37 @@ type ResolvedCredentials =
...
-    };
-  }
-
-  if (!fs.existsSync(projectPath) || !fs.statSync(projectPath).isDirectory()) {
-    return {
-      success: false,
-      error: `project_path "${projectPath}" does not exist or is not a directory`,
-    };
-  }
+  // Validation: ensure projectPath is a valid directory within the current working directory, with no path traversal
```
Comment (wdaimee) [inline src/utils/env-credentials.ts]: We should be able to pass in a project path outside the current working directory or MCP clients like Claude Desktop and Windsurf won't work as they aren't in an IDE environment. This comment will need to be updated.

File: src/utils/onboarding.ts (hunk 6 above)
Comment (wdaimee) [reply src/utils/onboarding.ts:37]: I am concerned that having a PROJECT_MARKERS array doesn't capture all the possible files for supported and unsupported frameworks. This tool should still run for unsupported frameworks like .NET, it uses the fallback logic instead of fetching the quickstart CDN spec. There seems to be some missing files in here, such as a `Package.swift` for Swift projects. Is there another solution we can implement that doesn't involve us having to keep this project markers array?

File: src/utils/env-credentials.ts (hunk 9 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts]: Removed this if statement logic

File: src/utils/env-credentials.ts (hunk 10 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts]: Updated this comment as well

File: src/utils/env-credentials.ts (hunk 5 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts:389]: I agree! I was also pondering whether to reduce the time further, I think 30 secs is a good balance for better UX as well. Updated it to 30 seconds.

File: test/utils/env-credentials.test.ts [hunk 11]
```diff
@@ -120,33 +123,87 @@ beforeEach(() => {
 });
 
 describe('resolveAndWriteCredentials — project path validation', () => {
-  it('returns error when project_path is not absolute', async () => {
```
Comment (mchang16-auth) [inline test/utils/env-credentials.test.ts:123]: Removed this test case since we are not using `isAbsolute()` anymore (`path.resolve()` makes the path absolute within `validateProjectPath()`)

File: src/utils/credentials-writer.ts [hunk 12]
```diff
@@ -63,8 +81,22 @@ export async function writeCredentialsToEnv(
...
+    } catch {
+      throw new Error('Failed to read existing env file');
+    }
+    let updatedLines: string;
+    try {
+      updatedLines = withTimeout(
+        () => commentOutConflictingKeys(existingContent, incomingKeys),
+        ENV_PARSE_TIMEOUT_MS,
+        'Env file processing'
+      );
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts]: Here we use a timeout to ensure when parsing the env file it does not catch when reading a potentially large file (which should not be realistic anyway for a env file)

File: src/utils/credentials-writer.ts [hunk 13]
```diff
@@ -51,10 +55,24 @@ export async function writeCredentialsToEnv(
...
+    }
+    const stat = fs.statSync(resolvedPath);
+    if (!stat.isFile()) {
+      throw new Error('Security error: env file path is not a regular file');
+    }
+    if (stat.size > MAX_ENV_FILE_SIZE_BYTES) {
+      throw new Error(
+        `Security error: env file exceeds maximum allowed size of ${MAX_ENV_FILE_SIZE_BYTES / 1024 / 1024} MB. ` +
+        'The file may be corrupted or malicious. Remove or truncate it before retrying.'
+      );
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:81]: This covers https://auth0team.atlassian.net/browse/SEC-32494 to prevent a large .env file causing the MCP to hang

File: src/utils/credentials-writer.ts [hunk 14]
```diff
@@ -84,15 +107,31 @@ export async function writeCredentialsToEnv(
...
+  // throw if chmod fails or if the resulting on-disk mode doesn't match.
+  // Windows has no meaningful chmod, so skip verification there.
+  if (process.platform !== 'win32') {
+    fs.chmodSync(envFile, CREDENTIAL_FILE_MODE);
+    const actualMode = fs.statSync(envFile).mode & 0o777;
+    if (actualMode !== CREDENTIAL_FILE_MODE) {
+      throw new Error(
+        `Security error: expected file mode ${CREDENTIAL_FILE_MODE.toString(8)}, got ${actualMode.toString(8)}`
+      );
+    }
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:138]: This covers https://auth0team.atlassian.net/browse/SEC-32495 and prevents access if the file has a non 600 permission

File: src/utils/env-credentials.ts [hunk 15]
```diff
@@ -55,26 +57,49 @@ type ResolvedCredentials =
...
+]);
+
+function isLikelyWebServedDirectory(resolvedDir: string): boolean {
+  const segment = path.basename(resolvedDir).toLowerCase();
+  if (WEB_SERVED_SEGMENT_NAMES.has(segment)) return true;
+  return WEB_SERVER_ROOTS.some(
+    root => resolvedDir === root || resolvedDir.startsWith(root + path.sep)
+  );
+}
+
```
Comment (mchang16-auth) [inline src/utils/env-credentials.ts:68]: This function covers https://auth0team.atlassian.net/browse/SEC-32459 and warns the user to double check their .env file is not made public on the web server. The method is called in `src/utils/env-credentials.ts`

File: src/server.ts [hunk 16]
```diff
@@ -111,6 +115,17 @@ export async function startServer(options?: ServerOptions) {
...
+          const { allowed, retryAfterMs } = rateLimiter(toolName);
+          if (!allowed) {
+            const retryAfterSec = Math.ceil(retryAfterMs / 1000);
+            throw new Error(
+              `Rate limit exceeded for ${toolName}: maximum 5 calls per 60 seconds. ` +
+              `Try again in ${retryAfterSec} second(s).`
+            );
+          }
+        }
+
```
Comment (mchang16-auth) [inline src/server.ts:129]: This method covers https://auth0team.atlassian.net/browse/SEC-32517 and imposes a sliding window rate limiting (checks the last 60 secs for calls made)

File: src/utils/credentials-writer.ts [hunk 17]
```diff
@@ -51,19 +55,37 @@ export async function writeCredentialsToEnv(
...
+    if (stat.size > MAX_ENV_FILE_SIZE_BYTES) {
+      throw new Error(
+        `Security error: env file exceeds maximum allowed size of ${MAX_ENV_FILE_SIZE_BYTES / 1024 / 1024} MB. ` +
+        'The file may be corrupted or malicious. Remove or truncate it before retrying.'
+      );
+    }
   }
-
-  const fileExisted = fs.existsSync(envFile);
```
Comment (mchang16-auth) [inline src/utils/credentials-writer.ts:60]: fileExisted is moved upward (line 61) to avoid a TOCTOU (time-of-check to time-of-use) window between the security checks and the subsequent read/write.

File: src/utils/onboarding.ts (hunk 6 above)
Comment (mchang16-auth) [reply src/utils/onboarding.ts:37]: Ended up changing it to a dis-allow list instead.

File: src/utils/rate-limiter.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+    if (timestamps.length >= maxCalls) {
+      return { allowed: false, retryAfterMs: timestamps[0] + windowMs - now };
+    }
+
+    timestamps.push(now);
+    calls.set(key, timestamps);
+    return { allowed: true, retryAfterMs: 0 };
+  };
+}
```
Comment (mchang16-auth) [inline src/utils/rate-limiter.ts:22]: `createRateLimiter(5, 60_000)` is called once at startup and returns a function. Each time you call that function with a key (e.g. '`auth0_save_credentials_to_file`'), it either allows or denies the call:

- It keeps a list of timestamps for each key, pruning any older than 60 seconds on every check
- If fewer than 5 timestamps remain after pruning, the call is allowed and the current timestamp is recorded
- If 5 or more remain, the call is denied and retryAfterMs tells the caller how long until the oldest timestamp expires out of the window

File: src/utils/credentials-writer.ts [hunk 19]
```diff
@@ -51,19 +55,37 @@ export async function writeCredentialsToEnv(
...
   const incomingKeys = new Set(Object.keys(credentials));
 
   let content: string;
 
   if (fileExisted) {
```
Comment (wdaimee) [inline src/utils/credentials-writer.ts]: This if block is now repeated (same as line 62) should we move the below logic into there or do you feel these are separate?

File: src/utils/env-credentials.ts [hunk 20]
```diff
@@ -312,3 +383,96 @@ async function buildFallbackCredentials(
...
+  try {
+    const guardStat = fs.statSync(statePath);
+    if (!guardStat.isFile() || guardStat.size > MAX_WRITE_GUARD_SIZE_BYTES) return null;
+    const state: WriteGuardState = JSON.parse(fs.readFileSync(statePath, 'utf-8'));
+    const elapsed = Date.now() - new Date(state.lastWrittenAt).getTime();
+    if (elapsed < WRITE_GUARD_WINDOW_MS) {
+      const overlap = incomingKeys.filter((k) => state.keysWritten.includes(k));
+      if (overlap.length > 0) {
+        const minutesAgo = Math.round(elapsed / 60_000);
+        const timeAgo = minutesAgo < 1 ? 'less than a minute' : `${minutesAgo} minute(s)`;
```
Comment (wdaimee) [inline src/utils/env-credentials.ts]: I think we need to update this to 30s, correct?

File: src/utils/env-credentials.ts (hunk 3 above)
Comment (wdaimee) [reply src/utils/env-credentials.ts:85]: In the tool definitions, we are stating that the project path is an absolute path. I am okay doing this and double-checking, however, we do kind lose the preventative behaviour of stopping the user from providing a relative path. I am okay either way, but it's a bit inconsistent with the schema definitions:

- applications.ts:365 — 'Absolute path to the project root directory. The env file is written here.'
  - quickstarts.ts:45 — 'Absolute path to the project directory. Used for .env file check and project config port 
  detection.'

File: src/utils/env-credentials.ts [hunk 21]
```diff
@@ -387,7 +382,7 @@ async function buildFallbackCredentials(
 // ── Write guard ──────────────────────────────────────────────────────────────
 // Prevents accidental double-writes within a 30-second window by persisting the
 // last-written keys and timestamp to .auth0-mcp-state.json in the project directory.
-// Can be bypassed with force: true for intentional overwrites.
+// Can be bypassed by deleting .auth0-mcp-state.json from the project directory.
```
Comment (wdaimee) [inline src/utils/env-credentials.ts]: I would remove this comment. I don't think we need to relay the message on how the write guard can be bypassed.

File: src/utils/env-credentials.ts [hunk 22]
```diff
@@ -413,11 +408,10 @@ function checkWriteGuard(projectPath: string, incomingKeys: string[]): string |
...
       const overlap = incomingKeys.filter((k) => state.keysWritten.includes(k));
       if (overlap.length > 0) {
-        const minutesAgo = Math.round(elapsed / 60_000);
-        const timeAgo = minutesAgo < 1 ? 'less than a minute' : `${minutesAgo} minute(s)`;
+        const secondsAgo = Math.round(elapsed / 1000);
         return (
-          `Credentials were already written to this project ${timeAgo} ago ` +
-          `(keys: ${overlap.join(', ')}). Pass force: true to overwrite.`
+          `Credentials were already written to this project ${secondsAgo} second(s) ago ` +
+          `(keys: ${overlap.join(', ')}). To write again, delete ${WRITE_GUARD_FILE} from the project directory and retry, or wait ${Math.ceil((WRITE_GUARD_WINDOW_MS - elapsed) / 1000)} second(s) for the window to expire.`
```
Comment (wdaimee) [inline src/utils/env-credentials.ts]: Do we want to give them this message about deleting the write guard file? I think the only option should be to let the user wait before trying to write to the file again.

File: src/utils/env-credentials.ts (hunk 21 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts]: Yeah I agree, I was also thinking it was a bit iffy. I'll remove the messaging

File: src/utils/env-credentials.ts (hunk 20 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts]: Updated the messaging to seconds

File: src/utils/credentials-writer.ts (hunk 19 above)
Comment (mchang16-auth) [reply src/utils/credentials-writer.ts]: Good catch, I'll merge the two if statements

**Review by wdaimee — DISMISSED** [review]
LGTM

File: src/server.ts (hunk 16 above)
Comment (kushalshit27) [reply src/server.ts:129]: Does the rate limiter apply to `dry_run` calls at the server layer, before the handler knows the call is non-writing?

File: src/utils/credentials-writer.ts (hunk 7 above)
Comment (kushalshit27) [reply src/utils/credentials-writer.ts:72]: The realpathSync check only runs inside if (fileExisted), so it only protects against a symlinked file. It doesn't help when the directory itself is a symlink.

File: src/utils/env-credentials.ts [hunk 23]
```diff
@@ -55,26 +56,43 @@ type ResolvedCredentials =
...
+  'public', 'dist', 'build', 'static', 'www', 'wwwroot', 'html', 'assets', 'out',
+]);
+
+// Detects whether a directory is likely web-served based on its final path segment.
+// Absolute web server roots (/srv, /var/www, etc.) are hard-blocked by hasProjectMarker.
+function isLikelyWebServedDirectory(resolvedDir: string): boolean {
+  return WEB_SERVED_SEGMENT_NAMES.has(path.basename(resolvedDir).toLowerCase());
+}
+
+function validateProjectPath(projectPath: string): string | null {
```
Comment (kushalshit27) [inline src/utils/env-credentials.ts:69]: Looks like the path.isAbsolute() check got dropped here. The new traversal check only catches paths within them. It prevents the resolved directory from changing depending on the server's CWD at call time, which matters in containers or multi-user setups.

Comment (mchang16-auth) [reply src/utils/env-credentials.ts:69]: @kushalshit27 thanks for pointing that out, I've gone ahead and added the isAbsolute() check back in, along with some refactoring that comes with path.resolve() not being needed then anymore

File: src/utils/env-credentials.ts (hunk 3 above)
Comment (mchang16-auth) [reply src/utils/env-credentials.ts:85]: @wdaimee I added back the `isAbsolute()` check since like what you mentioned and what @kushalshit27 the check is needed

File: src/utils/credentials-writer.ts (hunk 7 above)
Comment (mchang16-auth) [reply src/utils/credentials-writer.ts:72]: @kushalshit27 thanks for pointing that out, I added line 54 to have a symlink check for the directory specifically

File: src/server.ts (hunk 16 above)
Comment (mchang16-auth) [reply src/server.ts:129]: The `dry_run` would get rate limited, which I don't think we should have since no writes are actually happening. I've gone and removed the rate limited if `dry_run` is `true` (line 119)

**Review by kushalshit27 — APPROVED** [review]
LGTM!

**Review by wdaimee — APPROVED** [review]
LGTM
